Code/Framework/AzCore/AzCore/DOM/DomValue.cpp

```diff
@@ -0,0 +1,1241 @@
+                    {
+                        const Array::ContainerType& arrayContainer = GetArrayInternal();
+                        for (const Value& entry : arrayContainer)
+                        {
+                            result = entry.Accept(visitor, copyStrings);
+                            if (!result.IsSuccess())
+                            {
+                                return;
+                            }
+                        }
+                        result = visitor.EndArray(arrayContainer.size());
+                    }
+                }
+                else if constexpr (AZStd::is_same_v<Alternative, NodePtr>)
+                {
+                    const Node& node = *AZStd::get<NodePtr>(m_value);
+                    result = visitor.StartNode(node.GetName());
+                    if (result.IsSuccess())
+                    {
+                        const Object::ContainerType& object = GetObjectInternal();
+                        for (const Object::EntryType& entry : object)
+                        {
+                            result = visitor.Key(entry.first);
+                            if (!result.IsSuccess())
+                            {
+                                return;
+                            }
+                            result = entry.second.Accept(visitor, copyStrings);
+                            if (!result.IsSuccess())
+                            {
+                                return;
+                            }
+                        }
+
+                        const Array::ContainerType& arrayContainer = GetArrayInternal();
+                        for (const Value& entry : arrayContainer)
+                        {
+                            result = entry.Accept(visitor, copyStrings);
+                            if (!result.IsSuccess())
+                            {
+                                return;
+                            }
+                        }
+
+                        result = visitor.EndNode(object.size(), arrayContainer.size());
+                    }
+                }
+                else if constexpr (AZStd::is_same_v<Alternative, AZStd::any>)
+                {
+                    result = visitor.OpaqueValue(*arg);
+                }
+            },
+            m_value);
+        return result;
+    }
+
+    AZStd::unique_ptr<Visitor> Value::GetWriteHandler()
+    {
+        return AZStd::make_unique<ValueWriter>(*this);
+    }
+
+    bool Value::DeepCompareIsEqual(const Value& other) const
+    {
+        if (IsString() && other.IsString())
+        {
+            // If we both hold the same ref counted string we don't need to do a full comparison
+            if (AZStd::holds_alternative<AZStd::shared_ptr<const AZStd::string>>(m_value) && m_value == other.m_value)
+            {
+                return true;
+            }
+            return GetString() == other.GetString();
+        }
```

Comment:
Can this be done inside the `AZStd::visit`?

Reply:
It's more straightforward to do the check up-front so I can do a strict index comparison, but I'm open to revisiting this.